Small fixes

bbenligiray opened the pull request:
A bunch of small fixes:

- In Api3Voting, `EPOCH_LENGTH` is already assigned to `voteTime` in `initialize()` and since `EPOCH_LENGTH` and `voteTime` don't change, we can use `voteTime` instead of fetching `EPOCH_LENGTH` from the pool contract. It will just be slightly cheaper.

- Typecasts at checkpoint arrays were `assert()`ed but the ones at reward records weren't, so added it for the sake of completeness.

- `totalStake` is added to the reward records. This is actually not used in anywhere in the contract, but the problem I wanted to solve is that we don't keep any history of total staked amount on-chain for someone to build an external contract that uses that somehow. It just makes reward payments a bit more expensive so felt like it's worth it.

- Delegation and delegation update events being the same was confusing, so created another one.

- Small formatting and comment stuff

acenolaza commented on packages/convenience/contracts/Convenience.sol:
```diff
@@ -230,6 +278,13 @@ contract Convenience is Ownable  {
     /// @param userAddress User address
     /// @param voteIds Array of vote IDs for which data will be retrieved
+    /// @return executed If the vote has been executed
+    /// @return yea Total voting power voted for "For"
```
"In favor of" maybe better? "pro" is another alternative

bbenligiray replied:
I wanted to use the exact same word from the dashboard. The voting modal options are For and Against there.

acenolaza commented on packages/pool/contracts/RewardUtils.sol:
```diff
@@ -22,10 +22,13 @@ abstract contract RewardUtils is GetterUtils, IRewardUtils {
                     amount: uint224(rewardAmount),
-                    totalSharesThen: totalShares()
+                    totalSharesThen: totalShares(),
+                    totalStakeThen: totalStake
```
Another alternative could be publish a "RewardMinted" event and keep the totalStakeThen there for anyone to query.

bbenligiray replied:
This information can already be derived from the events, this is for the data to be on-chain, usable by other contracts.
Say one wants to airdrop tokens to API3 stakers 1:1 for each API3 token they have staked at a point in time. We normally never record historical `totalStake` so this can only be done for the current block. This provides such use-cases with milestones they can use for that kind of stuff. Not really necessary, but doesn't hurt either.
I'll still add it as a field to `RewardMinted` though, that will make it easier to calculate `totalStake` off-chain at a point in time easier.

acenolaza reviewed (approved):
LGTM 👍🏻 
Just added a few minor comments that might be useful.

bbenligiray commented on packages/api3-voting/contracts/Api3Voting.sol:
```diff
@@ -264,7 +264,7 @@ contract Api3Voting is IForwarder, AragonApp {
             uint256 lastProposalTimestamp
             ) = api3Pool.getUser(msg.sender);
-        require(lastProposalTimestamp.add(api3Pool.EPOCH_LENGTH()) < now, "API3_HIT_PROPOSAL_COOLDOWN");
+        require(lastProposalTimestamp.add(voteTime) < now, "API3_HIT_PROPOSAL_COOLDOWN");
```
I tried, turns out `immutable` was introduced in Solidity 0.6.5 :smile:

bbenligiray commented:
I added `totalStake` to the `RewardMinted` event as you've mentioned. Then I also added `totalShares` to `Staked` and `ScheduledUnstake`as this was something recommended by Team Omega informally. Then I updated the events to log all major state variables (that change) to make it easier to index these in the future. This doesn't affect how the current dashboard works at all.

acenolaza commented on packages/pool/contracts/TimelockUtils.sol:
```diff
@@ -46,13 +46,15 @@ abstract contract TimelockUtils is ClaimUtils, ITimelockUtils {
         // Should never return false because the API3 token uses the
         // OpenZeppelin implementation
         assert(api3Token.transferFrom(source, address(this), amount));
         emit DepositedByTimelockManager(
```
Should we also add source to the event just for completeness and consistency since we are adding all inputs on other events?

bbenligiray replied:
`source` will always be `timelockManager` (which is `immutable`) due to the awkward way TimelockManager.sol was implemented, so not really needed (could be done for consistency but meh)

acenolaza commented on packages/pool/contracts/TimelockUtils.sol:
```diff
@@ -118,7 +122,9 @@ abstract contract TimelockUtils is ClaimUtils, ITimelockUtils {
             userAddress,
```
same here in this event

bbenligiray replied:
`source` is also only `timelockManager`

acenolaza reviewed (approved):
LGTM again. I was just wondering, should we add "To" to variables like `userSharesUpdate` (`userSharesToUpdate`) for consistency with other local variables (i.e. delegatedToUpdate)?

bbenligiray commented:
> LGTM again. I was just wondering, should we add "To" to variables like `userSharesUpdate` (`userSharesToUpdate`) for consistency with other local variables (i.e. delegatedToUpdate)?

`delegatedToUpdate` is called that because the value it updates is called `delegatedTo`, so this is consistent (it's just that `delegatedTo` is weird for a variable name)
